tests/loaders/mod.rs

```diff
@@ -25,24 +25,57 @@ pub fn load_lang(file: &str) -> DefinitionBook {
   parse_lang(&code)
 }
 
-// For every pair in the map, replaces all matches of a string the other string
+// For every pair in the map, replaces all matches of a string with the other string
 pub fn replace_template(mut code: String, map: &[(&str, &str)]) -> String {
   for (from, to) in map {
     code = code.replace(from, to);
   }
   code
 }
 
+#[track_caller]
+pub fn hvm_lang_readback(net: &Net, book: &DefinitionBook, id_map: HashMap<run::Val, DefId>) -> String {
+  let net = hvm_lang::net::hvmc_to_net(net, &|val| id_map[&val]).unwrap();
+  let (res_term, valid_readback) = hvm_lang::term::net_to_term::net_to_term_non_linear(&net, book);
+  assert!(valid_readback);
+
+  res_term.to_string(&book.def_names)
+}
+
+pub fn hvm_lang_normal(book: &mut DefinitionBook, size: usize) -> (run::Net, Net, HashMap<run::Val, DefId>) {
+  let (compiled, id_map) = hvm_lang::compile_book(book).unwrap();
+  let (root, res_lnet) = normal(compiled, size);
+  (root, res_lnet, id_map)
+}
+
+#[allow(unused_variables)]
 pub fn normal(book: Book, size: usize) -> (run::Net, Net) {
   let book = book_to_runtime(&book);
-  let mut rnet = run::Net::new(size);
-  rnet.boot(name_to_val("main"));
-  rnet.normal(&book);
+
+  let rnet = {
+    #[cfg(not(feature = "cuda"))]
+    {
+      normal_cpu(book, size)
+    }
+    #[cfg(feature = "cuda")]
+    {
+      normal_gpu(book)
+    }
+  };
 
   let net = net_from_runtime(&rnet);
   (rnet, net)
 }
 
-pub fn hvm_lang_normal(book: DefinitionBook, size: usize) -> (Term, DefNames, hvm_lang::RunInfo) {
-  hvm_lang::run_book(book, size).unwrap()
+pub fn normal_cpu(book: run::Book, size: usize) -> run::Net {
```

Comment:
do you plan on calling either `normal_cpu` or `normal_gpu` outside this module? if no(and I think no, as well) make them private, or even better, put their definition inside the `normal` function